src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,411 @@
+        public long CalculateIntrinsicGas(ContractEntry transaction, IReleaseSpec releaseSpec)
+        {
+            long result = GasCostOf.Transaction;
+            long txDataNonZeroGasCost = releaseSpec.IsEip2028Enabled ? GasCostOf.TxDataNonZeroEip2028 : GasCostOf.TxDataNonZero;
+
+            if (transaction.Data != null)
+            {
+                for (int i = 0; i < transaction.Data.Length; i++)
+                {
+                    result += transaction.Data[i] == 0 ? GasCostOf.TxDataZero : txDataNonZeroGasCost;
+                }
+            }
+
+            if (transaction.IsValidDeploymentEntry && releaseSpec.IsEip2Enabled)
+            {
+                result += GasCostOf.TxCreate;
+            }
+
+            return result;
+        }
+        
+        [Todo(Improve.MissingFunctionality, "We need to agree on delta to state mapping details")]
+        private StateUpdate ToStateUpdate(Delta delta)
+        {
+            StateUpdate result = new StateUpdate
+            {
+                Difficulty = 1,
+                Number = 1,
+                Timestamp = (UInt256) delta.TimeStamp.Seconds,
+                GasLimit = delta.GasLimit,
+                /* here we can read coinbase entries from the delta
+                   but we need to decide how to split fees and which one to pick for the KVM */
+                GasBeneficiary = Address.Zero,
+                GasUsed = 0L
+            };
+
+            return result;
+        }
+
+        [Todo("Wider work needed to split calls and execution properly")]
+        public void CallAndRestore(Delta stateUpdate, ITxTracer txTracer, Address recipientOverride = null) { Execute(stateUpdate, txTracer, true, recipientOverride); }
+
+        [Todo("After delta is executed we should validate the state root and if it is not as expected we should revert all the changes.")]
+        public void Execute(Delta stateUpdate, ITxTracer txTracer, Address recipientOverride = null) { Execute(stateUpdate, txTracer, false, recipientOverride); }
+
+        Address GetAccountAddress(ByteString publicKeyByteString)
+        {
+            if (publicKeyByteString == null || publicKeyByteString.IsEmpty)
+            {
+                return null;
+            }
+
+            var publicKey = _cryptoContext.GetPublicKeyFromBytes(publicKeyByteString.ToByteArray());
+            return publicKey.ToKvmAddress();
+        }
+
+        private void QuickFail(ContractEntry entry, StateUpdate stateUpdate, ITxTracer txTracer, bool readOnly)
+        {
+            long gasLimit = entry.GasLimit;
+            var receiver = GetAccountAddress(entry.Base.ReceiverPublicKey);
+            var sender = GetAccountAddress(entry.Base.SenderPublicKey);
+            stateUpdate.GasUsed += gasLimit;
+            Address recipient = receiver ?? Address.OfContract(sender, _stateProvider.GetNonce(sender));
+            if (txTracer.IsTracingReceipt) txTracer.MarkAsFailed(recipient, gasLimit, Bytes.Empty, "invalid");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4350015)